torch/csrc/jit/script/concrete_module_type.h

```diff
@@ -256,11 +184,92 @@ class VISIBILITY_HIDDEN ConcreteModuleType {
   // The original `nn.Module` class that we derived this ScriptModule from.
   py::object pyClass_;
 
-  // The JIT type derived from this ConcreteModuleType.
-  ClassTypePtr jitType_ = nullptr;
   // NOTE: If you ever add any more state to this struct, you need to make sure
-  // operator== still makes sense! The only field that can be excluded from it
-  // is `jitType_`.
+  // operator== still makes sense!
+};
+
+// Represents a concrete type during in the process for construction. We use
+// this to decide whether we can share types between modules.
+class VISIBILITY_HIDDEN RawConcreteModuleType {
```

Comment:
If ConcreteModuleType was a friend of its builder, then we would not need ConcreteModuleTypeData and its data_ indirections. I don't think that changes its safety properties as long as the builder returns the built type by value to cause it to create a copy.